scalyr_agent/log_processing.py

```diff
@@ -1701,6 +1737,9 @@ def perform_processing(self, add_events_request, current_time=None):
             self.__lock.acquire()
             self.__total_bytes_being_processed = bytes_copied
             self.__total_bytes_pending = self.__log_file_iterator.available
+
+            # also keep track of the amount of bytes read this time
+            self.__prev_bytes_read = bytes_read
```

Comment:
I dislike that github prevents me from add comments to lines that you didn't change.  Argh.  So, I'll add my comments here for this method.

You should update `_is_closed` in this method as part of the completion callback.  You can see we already update `_is_closed` there for the other conditions.  I always view that the completion back is when _commit_ changes to the `LogProcessor`.  So, that's when we should update `_is_closed`.

For the logic, I think we should set `_is_closed` to `True` if `_close_when_caught_up` was true BEFORE the call above to line 1598 when we invoke `self.__log_file_iterator.scan_for_new_bytes`.  This prevents a subtle race condition.  We need to only consider us being able to close if someone has asked us to close when caught up before we noticed there are no new bytes to be read.  So, that means we should just capture a local copy of `_close_when_caught_up` up there and then refer to it in the completion callback.